goldenmaster/apigear/mqtt/private/mqttcwrapper.cpp

```diff
@@ -0,0 +1,587 @@
+    MQTTAsync_disconnectOptions disconn_opts = MQTTAsync_disconnectOptions_initializer5;
+    disconn_opts.onSuccess5 = ::onDisconnected;
+    disconn_opts.context = new genericContext{getPtr()};
+    disconn_opts.timeout = 10;
+    MQTTAsync_disconnect(*m_client.get(), &disconn_opts);
+    m_onConnectionStatusChangedCallbacksMutex.lock();
+    auto onConnectionStatusChangedCallbacks(std::move(m_onConnectionStatusChangedCallbacks));
+    m_onConnectionStatusChangedCallbacks.clear();
+    m_onConnectionStatusChangedCallbacksMutex.unlock();
+    for (auto& callback: onConnectionStatusChangedCallbacks){
+        callback.second(false);
+    }
+}
+
+void CWrapper::onConnected()
+{
+    m_connected = true;
+    AG_LOG_DEBUG("socket connected");
+    m_onConnectionStatusChangedCallbacksMutex.lock();
+    auto onConnectionStatusChangedCallbacks { m_onConnectionStatusChangedCallbacks };
+    m_onConnectionStatusChangedCallbacksMutex.unlock();
+    for (auto& callback: onConnectionStatusChangedCallbacks){
+        callback.second(true);
+    }
+    m_mainMQTTThread = std::thread(&CWrapper::run, this);
+}
+
+void CWrapper::resubscribeAllTopics()
+{
+    m_subscribedTopicsMutex.lock();
+    m_toBeSubscribedTopicsMutex.lock();
+    if(!m_subscribedTopics.empty())
+    {
+        m_subscribedTopics.insert(m_toBeSubscribedTopics.begin(),m_toBeSubscribedTopics.end());
+        m_toBeSubscribedTopics = m_subscribedTopics;
+        m_subscribedTopics.clear();
+    }
+    m_toBeSubscribedTopicsMutex.unlock();
+    m_subscribedTopicsMutex.unlock();
+}
+
+void CWrapper::onDisconnected()
+{
+    m_connected = false;
+    bool disconnectRequested = m_disconnectRequested;
+    m_disconnectRequested = false;
+    AG_LOG_DEBUG("socket disconnected");
+
+    // if we have not waited for our thread to finish, do it now
+    if (m_mainMQTTThread.joinable())
+    {
+        {
+            std::lock_guard<std::mutex> l{m_waitForSubscriptionChangesMutex};
+            m_waitForSubscriptionChanges = false;
+        }
+        m_synchronizeSubscriptionChanges.notify_one();
+        m_mainMQTTThread.join();
+    }
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+    // this function is called from within the MQTTAsync client
+    // therefore the client must be reset in a separate thread afterwards
+    std::thread([this, disconnectRequested](){
+        // use MQTT function to properly delete the client before resetting the unique ptr
+        MQTTAsync_destroy(m_client.get());
+        m_client.reset();
+
+        // reconnect if the connection was not dropped intentionally by us
+        if(!disconnectRequested)
+        {
+            // we need to re-subscribe to all topics on re-connection
+            resubscribeAllTopics();
+
+            connectToHost(m_serverUrl);
+        }
+    }).detach();
+}
+
+void CWrapper::handleTextMessage(const Message& message)
+{
+    AG_LOG_INFO("new msg: topic " + message.topic + " msg content: " + message.content);
```

Comment:
very heavy log on critical path, the string is always constructed

Reply:
done